src/main/cpp/src/decimal_utils.cu

```diff
@@ -0,0 +1,542 @@
+    }
+
+}
+
+// check that the divide is going to do the right thing
+void check_scale_divisor(int source_scale, int target_scale) {
+  CUDF_EXPECTS(source_scale <= target_scale, "source scale is larger than target scale");
+  int exponent = target_scale - source_scale;
+  CUDF_EXPECTS(exponent <= cuda::std::numeric_limits<__int128_t>::digits10, "divisor too big");
+}
+
+// This gets a little complicated. We are trying to find the precision of a large
+// number that could up to 256 bits.
+inline __device__ int precision10(chunked256 value) {
+    if (value.sign() < 0) {
+      // we wnat to do this on positive numbers
+      value.negate();
+    }
+    for (int i = 0; i <= 76; i++) {
+      chunked256 tmp = pow_ten(i);
+      if (tmp.gte_unsigned(value)) {
+        return i;
+      }
+    }
+    return -1;
+}
+
+// Functor to multiply two DECIMAL128 columns with rounding and overflow detection.
+struct dec128_multiplier : public thrust::unary_function<cudf::size_type, __int128_t> {
+  dec128_multiplier(bool *overflows, cudf::mutable_column_view const &product_view,
+                    cudf::column_view const &a_col, cudf::column_view const &b_col)
+      : overflows(overflows), a_data(a_col.data<__int128_t>()), b_data(b_col.data<__int128_t>()),
+        product_data(product_view.data<__int128_t>()),
+        a_scale(a_col.type().scale()), b_scale(b_col.type().scale()),
+        prod_scale(product_view.type().scale()) {}
+
+  __device__ __int128_t operator()(cudf::size_type i) const {
+    chunked256 const a(a_data[i]);
+    chunked256 const b(b_data[i]);
+
+    chunked256 product = multiply(a, b);
+
+    // Spark does some really odd things that I personally think are a bug
+    // https://issues.apache.org/jira/browse/SPARK-40129
+    // But to match Spark we need to first round the result to a precision of 38
+    // and this is specific to the value in the result of the multiply.
+    // Then we need to round the result to the final scale that we care about.
+    int dec_precision = precision10(product);
+    int first_div_precision = dec_precision - 38;
+
+    int mult_scale = a_scale + b_scale;
+    if (first_div_precision > 0) {
+      __int128_t first_div_scale_divisor = std::pow(10, first_div_precision);
+      divmod256 div_result = divide(product, first_div_scale_divisor);
+      product = div_result.quotient;
+
+      bool const needIncrement = (div_result.remainder < 0 ? -div_result.remainder : div_result.remainder) 
+          >= (first_div_scale_divisor >> 1);
+      int64_t const sign = product.sign();
+      int const roundIncrement = (needIncrement ? sign < 0 ? -1 : 1 : 0);
+      product.add(roundIncrement);
+      // a_scale and b_scale are negative. first_div_precision is not
+      mult_scale = a_scale + b_scale + first_div_precision;
+    }
+
+    int exponent = prod_scale - mult_scale;
+    if (exponent < 0) {
+      overflows[i] = true;
+      return;
+    }
+    __int128_t scale_divisor = std::pow(10, exponent);
```

Comment:
Seems like we could refactor to leverage some common code with `pow_ten` here, maybe a TODO?